File: ccgo/commands/clean.py
#
# Copyright 2024 zhlinh and ccgo Project Authors. All rights reserved.
# Use of this source code is governed by a MIT-style
# license that can be found at
#
# https://opensource.org/license/MIT
#
# The above copyright notice and this permission
# notice shall be included in all copies or
# substantial portions of the Software.

import os
import sys
import argparse
import shutil
from pathlib import Path

# setup path
# >>>>>>>>>>>>>>
SCRIPT_PATH = os.path.split(os.path.realpath(__file__))[0]
PROJECT_ROOT_PATH = os.path.dirname(SCRIPT_PATH)
sys.path.append(SCRIPT_PATH)
sys.path.append(PROJECT_ROOT_PATH)
PACKAGE_NAME = os.path.basename(SCRIPT_PATH)
# <<<<<<<<<<<<<
# import this project modules
try:
    from ccgo.utils.context.namespace import CliNameSpace
    from ccgo.utils.context.context import CliContext
    from ccgo.utils.context.command import CliCommand
    from ccgo.utils.cmd.cmd_util import exec_command
except ImportError:
    from utils.context.namespace import CliNameSpace
    from utils.context.context import CliContext
    from utils.context.command import CliCommand
    from utils.cmd.cmd_util import exec_command


class Clean(CliCommand):
    def description(self) -> str:
        return """
        This is a subcommand to clean build artifacts and caches.

        Cleans the following directories:
        - bin/                    # Output binaries
        - cmake_build/            # CMake build directory
        - android/build/          # Android Gradle build cache
        - android/.gradle/        # Android Gradle cache
        - ohos/build/             # OHOS build cache
        - ohos/.hvigor/           # OHOS hvigor cache
        - kmp/build/              # KMP build cache
        - examples/*/build/       # Example projects build cache

        Examples:
            ccgo clean              # Clean all build artifacts (with confirmation)
            ccgo clean android      # Clean only Android platform
            ccgo clean --dry-run    # Preview what will be cleaned
            ccgo clean -y           # Clean all without confirmation
            ccgo clean --native-only  # Clean only cmake_build directory
        """

    def get_target_list(self) -> list:
        return ["all", "android", "ios", "macos", "ohos", "kmp", "examples"]

    def cli(self) -> CliNameSpace:
        parser = argparse.ArgumentParser(
            prog="ccgo clean",
            formatter_class=argparse.RawDescriptionHelpFormatter,
            description=self.description(),
        )
        parser.add_argument(
            "target",
            nargs='?',
            default="all",
            type=str,
            choices=self.get_target_list(),
            help="Platform to clean (default: all)",
        )
        parser.add_argument(
            "--native-only",
            action="store_true",
            help="Clean only cmake_build directory (CMake native builds)",
        )
        parser.add_argument(
            "--dry-run",
            action="store_true",
            help="Show what would be cleaned without actually deleting",
        )
        parser.add_argument(
            "-y", "--yes",
            action="store_true",
            help="Skip confirmation prompts",
        )
        module_name = os.path.splitext(os.path.basename(__file__))[0]
        input_argv = [x for x in sys.argv[1:] if x != module_name]
        args, unknown = parser.parse_known_args(input_argv)
        return args

    def exec(self, context: CliContext, args: CliNameSpace):
        print("Cleaning build artifacts and caches...\n")

        # Get current working directory (project directory)
        try:
            project_dir = os.getcwd()
        except (OSError, FileNotFoundError) as e:
            # If current directory was deleted, try to use PWD environment variable
            project_dir = os.environ.get('PWD')
            if not project_dir or not os.path.exists(project_dir):
                print(f"ERROR: Current working directory no longer exists: {e}")
                print("Please navigate to your project directory and try again.")
                sys.exit(1)
            # Try to change to the saved directory
            try:
                os.chdir(project_dir)
            except (OSError, FileNotFoundError):
                print(f"ERROR: Cannot access project directory: {project_dir}")
                sys.exit(1)

        # Check if CCGO.toml exists to verify we're in a CCGO project
        config_path = None
        for subdir in os.listdir(project_dir):
            subdir_path = os.path.join(project_dir, subdir)
            if not os.path.isdir(subdir_path):
                continue
            potential_config = os.path.join(subdir_path, "CCGO.toml")
            if os.path.isfile(potential_config):
                config_path = potential_config
                project_subdir = subdir_path
                break

        # If not found in subdirectory, check current directory
        if not config_path:
            if os.path.isfile(os.path.join(project_dir, "CCGO.toml")):
                config_path = os.path.join(project_dir, "CCGO.toml")
                project_subdir = project_dir
            else:
                print("⚠️  Warning: CCGO.toml not found. May not be in a CCGO project directory.")
                print("   Continuing anyway...\n")
                project_subdir = project_dir

        cleaner = ProjectCleaner(project_subdir, dry_run=args.dry_run, skip_confirm=args.yes)

        # Determine what to clean based on arguments
        if args.native_only:
            # Only clean cmake_build directory
            cleaner.clean_cmake()
        elif args.target == "all":
            # Clean all platforms
            cleaner.clean_all()
        else:
            # Clean specific platform
            cleaner.clean_platform(args.target)

        cleaner.print_summary()


class ProjectCleaner:
    def __init__(self, project_dir, dry_run=False, skip_confirm=False):
        self.project_dir = project_dir
        self.dry_run = dry_run
        self.skip_confirm = skip_confirm
        self.cleaned_dirs = []
        self.cleaned_size = 0
        self.failed_dirs = []

    def get_dir_size(self, path):
        """Get total size of directory in bytes"""
        total_size = 0
        try:
            for dirpath, dirnames, filenames in os.walk(path):
                for filename in filenames:
                    filepath = os.path.join(dirpath, filename)
                    try:
                        total_size += os.path.getsize(filepath)
                    except (OSError, FileNotFoundError):
                        pass
        except (OSError, PermissionError):
            pass
        return total_size

    def format_size(self, size_bytes):
        """Format bytes to human-readable size"""
        for unit in ['B', 'KB', 'MB', 'GB']:
            if size_bytes < 1024.0:
                return f"{size_bytes:.2f} {unit}"
            size_bytes /= 1024.0
        return f"{size_bytes:.2f} TB"

    def remove_directory(self, dir_path, dir_name=None):
        """Remove a directory and track the result"""
        if not os.path.exists(dir_path):
            return False

        if not os.path.isdir(dir_path):
            return False

        size = self.get_dir_size(dir_path)
        display_name = dir_name or os.path.basename(dir_path)

        if self.dry_run:
            print(f"  [DRY RUN] Would remove: {display_name} ({self.format_size(size)})")
            return True

        try:
            shutil.rmtree(dir_path)
            self.cleaned_dirs.append(display_name)
            self.cleaned_size += size
            print(f"  ✅ Removed: {display_name} ({self.format_size(size)})")
            return True
        except Exception as e:
            self.failed_dirs.append((display_name, str(e)))
            print(f"  ❌ Failed to remove {display_name}: {e}")
            return False

    def confirm_clean(self, message):
        """Ask user for confirmation"""
        if self.skip_confirm:
            return True

        response = input(f"{message} (y/N): ").strip().lower()
        return response in ['y', 'yes']

    def clean_bin(self):
        """Clean bin directory"""
        print("\n" + "="*60)
        print("  Cleaning bin/ directory")
        print("="*60)

        bin_dir = os.path.join(self.project_dir, "bin")

        if not os.path.exists(bin_dir):
            print("  ℹ️  bin/ directory does not exist")
            return

        if not self.dry_run and not self.skip_confirm:
            if not self.confirm_clean("  Remove bin/ directory?"):
                print("  ⏭️  Skipped")
                return

        self.remove_directory(bin_dir, "bin/")

    def clean_cmake(self):
        """Clean cmake_build directory"""
        print("\n" + "="*60)
        print("  Cleaning cmake_build/ directory")
        print("="*60)

        cmake_dir = os.path.join(self.project_dir, "cmake_build")

        if not os.path.exists(cmake_dir):
            print("  ℹ️  cmake_build/ directory does not exist")
            return

        if not self.dry_run and not self.skip_confirm:
            if not self.confirm_clean("  Remove cmake_build/ directory?"):
                print("  ⏭️  Skipped")
                return

        self.remove_directory(cmake_dir, "cmake_build/")

    def clean_android(self):
        """Clean Android build caches"""
        print("\n" + "="*60)
        print("  Cleaning Android build caches")
        print("="*60)

        android_dir = os.path.join(self.project_dir, "android")

        if not os.path.exists(android_dir):
            print("  ℹ️  android/ directory does not exist")
            return

        # Clean android/build/
        build_dir = os.path.join(android_dir, "build")
        if os.path.exists(build_dir):
            self.remove_directory(build_dir, "android/build/")

        # Clean android/.gradle/
        gradle_dir = os.path.join(android_dir, ".gradle")
        if os.path.exists(gradle_dir):
            self.remove_directory(gradle_dir, "android/.gradle/")

        # Clean android/*/build/ for all subprojects
        for item in os.listdir(android_dir):
            item_path = os.path.join(android_dir, item)
            if os.path.isdir(item_path):
                subproject_build = os.path.join(item_path, "build")
                if os.path.exists(subproject_build):
                    self.remove_directory(subproject_build, f"android/{item}/build/")

    def clean_ios(self):
        """Clean iOS/macOS build caches"""
        print("\n" + "="*60)
        print("  Cleaning iOS/macOS build caches")
        print("="*60)

        # Clean Pods cache
        pods_dir = os.path.join(self.project_dir, "ios", "Pods")
        if os.path.exists(pods_dir):
            self.remove_directory(pods_dir, "ios/Pods/")

        # Clean DerivedData
        derived_data = os.path.join(self.project_dir, "ios", "DerivedData")
        if os.path.exists(derived_data):
            self.remove_directory(derived_data, "ios/DerivedData/")

        # Clean build directories
        ios_build = os.path.join(self.project_dir, "ios", "build")
        if os.path.exists(ios_build):
            self.remove_directory(ios_build, "ios/build/")

    def clean_macos(self):
        """Clean macOS build caches (same as iOS)"""
        print("\n" + "="*60)
        print("  Cleaning macOS build caches")
        print("="*60)

        # Clean macOS build if separate from iOS
        macos_build = os.path.join(self.project_dir, "macos", "build")
        if os.path.exists(macos_build):
            self.remove_directory(macos_build, "macos/build/")

    def clean_ohos(self):
        """Clean OHOS build caches"""
        print("\n" + "="*60)
        print("  Cleaning OHOS build caches")
        print("="*60)

        ohos_dir = os.path.join(self.project_dir, "ohos")

        if not os.path.exists(ohos_dir):
            print("  ℹ️  ohos/ directory does not exist")
            return

        # Clean ohos/build/
        build_dir = os.path.join(ohos_dir, "build")
        if os.path.exists(build_dir):
            self.remove_directory(build_dir, "ohos/build/")

        # Clean ohos/.hvigor/
        hvigor_dir = os.path.join(ohos_dir, ".hvigor")
        if os.path.exists(hvigor_dir):
            self.remove_directory(hvigor_dir, "ohos/.hvigor/")

        # Clean ohos/*/build/ for all subprojects
        for item in os.listdir(ohos_dir):
            item_path = os.path.join(ohos_dir, item)
            if os.path.isdir(item_path) and item not in ['.hvigor', 'build']:
                subproject_build = os.path.join(item_path, "build")
                if os.path.exists(subproject_build):
                    self.remove_directory(subproject_build, f"ohos/{item}/build/")

    def clean_kmp(self):
        """Clean KMP build caches"""
        print("\n" + "="*60)
        print("  Cleaning KMP build caches")
        print("="*60)

        kmp_dir = os.path.join(self.project_dir, "kmp")

        if not os.path.exists(kmp_dir):
            print("  ℹ️  kmp/ directory does not exist")
            return

        # Clean kmp/build/
        build_dir = os.path.join(kmp_dir, "build")
        if os.path.exists(build_dir):
            self.remove_directory(build_dir, "kmp/build/")

        # Clean kmp/.gradle/
        gradle_dir = os.path.join(kmp_dir, ".gradle")
        if os.path.exists(gradle_dir):
            self.remove_directory(gradle_dir, "kmp/.gradle/")

    def clean_examples(self):
        """Clean examples build caches"""
        print("\n" + "="*60)
        print("  Cleaning examples build caches")
        print("="*60)

        examples_dir = os.path.join(self.project_dir, "examples")

        if not os.path.exists(examples_dir):
            print("  ℹ️  examples/ directory does not exist")
            return

        # Find all build directories in examples
        for root, dirs, files in os.walk(examples_dir):
            for dir_name in dirs:
                if dir_name in ['build', '.gradle', '.hvigor']:
                    full_path = os.path.join(root, dir_name)
                    rel_path = os.path.relpath(full_path, self.project_dir)
                    self.remove_directory(full_path, rel_path)

    def clean_platform(self, platform):
        """Clean specific platform caches"""
        if platform == "android":
            self.clean_android()
        elif platform == "ios":
            self.clean_ios()
        elif platform == "macos":
            self.clean_macos()
        elif platform == "ohos":
            self.clean_ohos()
        elif platform == "kmp":
            self.clean_kmp()
        elif platform == "examples":
            self.clean_examples()

    def clean_all(self):
        """Clean all build artifacts and caches"""
        print("\n" + "="*60)
        print("  Cleaning ALL build artifacts and caches")
        print("="*60)

        if not self.dry_run and not self.skip_confirm:
            if not self.confirm_clean("\n⚠️  This will remove ALL build artifacts. Continue?"):
                print("  ⏭️  Aborted")
                return

        # Clean in order
        self.clean_bin()
        self.clean_cmake()
        self.clean_android()
        self.clean_ios()
        self.clean_macos()
        self.clean_ohos()
        self.clean_kmp()
        self.clean_examples()

        # Also clean root .gradle if exists
        root_gradle = os.path.join(self.project_dir, ".gradle")
        if os.path.exists(root_gradle):
            print("\n" + "="*60)
            print("  Cleaning root .gradle/")
            print("="*60)
            self.remove_directory(root_gradle, ".gradle/")

    def print_summary(self):
        """Print summary of cleaning operation"""
        print("\n" + "="*60)
        print("  Cleaning Summary")
        print("="*60)

        if self.dry_run:
            print("  [DRY RUN MODE - No files were actually deleted]")

        if self.cleaned_dirs:
            print(f"  ✅ Successfully cleaned {len(self.cleaned_dirs)} directories:")
            for dir_name in self.cleaned_dirs:
                print(f"     - {dir_name}")
            print(f"\n  💾 Total space freed: {self.format_size(self.cleaned_size)}")
        else:
            print("  ℹ️  No directories were cleaned")

        if self.failed_dirs:
            print(f"\n  ❌ Failed to clean {len(self.failed_dirs)} directories:")
            for dir_name, error in self.failed_dirs:
                print(f"     - {dir_name}: {error}")

        print("="*60 + "\n")

        if self.dry_run:
            print("💡 Tip: Run without --dry-run to actually delete the files")

File: src/commands/clean.rs
//! Clean command implementation

use anyhow::{Context, Result};
use clap::Args;
use std::fs;
use std::io::{self, Write};
use std::path::{Path, PathBuf};
use walkdir::WalkDir;

/// Clean build artifacts
#[derive(Args, Debug)]
pub struct CleanCommand {
    /// Target to clean (all, android, ios, macos, ohos, kmp, examples)
    #[arg(default_value = "all")]
    pub target: String,

    /// Show what would be deleted
    #[arg(long)]
    pub dry_run: bool,

    /// Skip confirmation prompt
    #[arg(short = 'y', long)]
    pub yes: bool,

    /// Clean only cmake_build directory
    #[arg(long)]
    pub native_only: bool,
}

struct ProjectCleaner {
    project_dir: PathBuf,
    dry_run: bool,
    skip_confirm: bool,
    cleaned_dirs: Vec<String>,
    cleaned_size: u64,
    failed_dirs: Vec<(String, String)>,
}

impl ProjectCleaner {
    fn new(project_dir: PathBuf, dry_run: bool, skip_confirm: bool) -> Self {
        Self {
            project_dir,
            dry_run,
            skip_confirm,
            cleaned_dirs: Vec::new(),
            cleaned_size: 0,
            failed_dirs: Vec::new(),
        }
    }

    fn get_dir_size(path: &Path) -> u64 {
        let mut total_size = 0u64;
        if let Ok(entries) = WalkDir::new(path).into_iter().collect::<Result<Vec<_>, _>>() {
            for entry in entries {
                if let Ok(metadata) = entry.metadata() {
                    if metadata.is_file() {
                        total_size += metadata.len();
                    }
                }
            }
        }
        total_size
    }

    fn format_size(size_bytes: u64) -> String {
        const UNITS: &[&str] = &["B", "KB", "MB", "GB", "TB"];
        let mut size = size_bytes as f64;
        let mut unit_idx = 0;

        while size >= 1024.0 && unit_idx < UNITS.len() - 1 {
            size /= 1024.0;
            unit_idx += 1;
        }

        format!("{:.2} {}", size, UNITS[unit_idx])
    }

    fn remove_directory(&mut self, dir_path: &Path, display_name: Option<&str>) -> bool {
        if !dir_path.exists() {
            return false;
        }

        if !dir_path.is_dir() {
            return false;
        }

        let size = Self::get_dir_size(dir_path);
        let name = display_name.unwrap_or_else(|| {
            dir_path
                .file_name()
                .and_then(|n| n.to_str())
                .unwrap_or("unknown")
        });

        if self.dry_run {
            println!("  [DRY RUN] Would remove: {} ({})", name, Self::format_size(size));
            return true;
        }

        match fs::remove_dir_all(dir_path) {
            Ok(_) => {
                self.cleaned_dirs.push(name.to_string());
                self.cleaned_size += size;
                println!("  ✅ Removed: {} ({})", name, Self::format_size(size));
                true
            }
            Err(e) => {
                self.failed_dirs.push((name.to_string(), e.to_string()));
                println!("  ❌ Failed to remove {}: {}", name, e);
                false
            }
        }
    }

    fn confirm_clean(&self, message: &str) -> bool {
        if self.skip_confirm {
            return true;
        }

        print!("{} (y/N): ", message);
        io::stdout().flush().unwrap();

        let mut input = String::new();
        io::stdin().read_line(&mut input).unwrap();
        let input = input.trim().to_lowercase();

        input == "y" || input == "yes"
    }

    fn clean_bin(&mut self) {
        println!("\n{}", "=".repeat(60));
        println!("  Cleaning bin/ directory");
        println!("{}", "=".repeat(60));

        let bin_dir = self.project_dir.join("bin");

        if !bin_dir.exists() {
            println!("  ℹ️  bin/ directory does not exist");
            return;
        }

        if !self.dry_run && !self.skip_confirm {
            if !self.confirm_clean("  Remove bin/ directory?") {
                println!("  ⏭️  Skipped");
                return;
            }
        }

        self.remove_directory(&bin_dir, Some("bin/"));
    }

    fn clean_cmake(&mut self) {
        println!("\n{}", "=".repeat(60));
        println!("  Cleaning cmake_build/ directory");
        println!("{}", "=".repeat(60));

        let cmake_dir = self.project_dir.join("cmake_build");

        if !cmake_dir.exists() {
            println!("  ℹ️  cmake_build/ directory does not exist");
            return;
        }

        if !self.dry_run && !self.skip_confirm {
            if !self.confirm_clean("  Remove cmake_build/ directory?") {
                println!("  ⏭️  Skipped");
                return;
            }
        }

        self.remove_directory(&cmake_dir, Some("cmake_build/"));
    }

    fn clean_android(&mut self) {
        println!("\n{}", "=".repeat(60));
        println!("  Cleaning Android build caches");
        println!("{}", "=".repeat(60));

        let android_dir = self.project_dir.join("android");

        if !android_dir.exists() {
            println!("  ℹ️  android/ directory does not exist");
            return;
        }

        // Clean android/build/
        let build_dir = android_dir.join("build");
        if build_dir.exists() {
            self.remove_directory(&build_dir, Some("android/build/"));
        }

        // Clean android/.gradle/
        let gradle_dir = android_dir.join(".gradle");
        if gradle_dir.exists() {
            self.remove_directory(&gradle_dir, Some("android/.gradle/"));
        }

        // Clean android/*/build/ for all subprojects
        if let Ok(entries) = fs::read_dir(&android_dir) {
            for entry in entries.flatten() {
                if let Ok(file_type) = entry.file_type() {
                    if file_type.is_dir() {
                        let subproject_build = entry.path().join("build");
                        if subproject_build.exists() {
                            let name = format!("android/{}/build/", entry.file_name().to_string_lossy());
                            self.remove_directory(&subproject_build, Some(&name));
                        }
                    }
                }
            }
        }
    }

    fn clean_ios(&mut self) {
        println!("\n{}", "=".repeat(60));
        println!("  Cleaning iOS/macOS build caches");
        println!("{}", "=".repeat(60));

        // Clean Pods cache
        let pods_dir = self.project_dir.join("ios/Pods");
        if pods_dir.exists() {
            self.remove_directory(&pods_dir, Some("ios/Pods/"));
        }

        // Clean DerivedData
        let derived_data = self.project_dir.join("ios/DerivedData");
        if derived_data.exists() {
            self.remove_directory(&derived_data, Some("ios/DerivedData/"));
        }

        // Clean build directories
        let ios_build = self.project_dir.join("ios/build");
        if ios_build.exists() {
            self.remove_directory(&ios_build, Some("ios/build/"));
        }
    }

    fn clean_macos(&mut self) {
        println!("\n{}", "=".repeat(60));
        println!("  Cleaning macOS build caches");
        println!("{}", "=".repeat(60));

        // Clean macOS build if separate from iOS
        let macos_build = self.project_dir.join("macos/build");
        if macos_build.exists() {
            self.remove_directory(&macos_build, Some("macos/build/"));
        }
    }

    fn clean_ohos(&mut self) {
        println!("\n{}", "=".repeat(60));
        println!("  Cleaning OHOS build caches");
        println!("{}", "=".repeat(60));

        let ohos_dir = self.project_dir.join("ohos");

        if !ohos_dir.exists() {
            println!("  ℹ️  ohos/ directory does not exist");
            return;
        }

        // Clean ohos/build/
        let build_dir = ohos_dir.join("build");
        if build_dir.exists() {
            self.remove_directory(&build_dir, Some("ohos/build/"));
        }

        // Clean ohos/.hvigor/
        let hvigor_dir = ohos_dir.join(".hvigor");
        if hvigor_dir.exists() {
            self.remove_directory(&hvigor_dir, Some("ohos/.hvigor/"));
        }

        // Clean ohos/*/build/ for all subprojects
        if let Ok(entries) = fs::read_dir(&ohos_dir) {
            for entry in entries.flatten() {
                if let Ok(file_type) = entry.file_type() {
                    if file_type.is_dir() {
                        let dir_name = entry.file_name();
                        let dir_name_str = dir_name.to_string_lossy();
                        if dir_name_str != ".hvigor" && dir_name_str != "build" {
                            let subproject_build = entry.path().join("build");
                            if subproject_build.exists() {
                                let name = format!("ohos/{}/build/", dir_name_str);
                                self.remove_directory(&subproject_build, Some(&name));
                            }
                        }
                    }
                }
            }
        }
    }

    fn clean_kmp(&mut self) {
        println!("\n{}", "=".repeat(60));
        println!("  Cleaning KMP build caches");
        println!("{}", "=".repeat(60));

        let kmp_dir = self.project_dir.join("kmp");

        if !kmp_dir.exists() {
            println!("  ℹ️  kmp/ directory does not exist");
            return;
        }

        // Clean kmp/build/
        let build_dir = kmp_dir.join("build");
        if build_dir.exists() {
            self.remove_directory(&build_dir, Some("kmp/build/"));
        }

        // Clean kmp/.gradle/
        let gradle_dir = kmp_dir.join(".gradle");
        if gradle_dir.exists() {
            self.remove_directory(&gradle_dir, Some("kmp/.gradle/"));
        }
    }

    fn clean_examples(&mut self) {
        println!("\n{}", "=".repeat(60));
        println!("  Cleaning examples build caches");
        println!("{}", "=".repeat(60));

        let examples_dir = self.project_dir.join("examples");

        if !examples_dir.exists() {
            println!("  ℹ️  examples/ directory does not exist");
            return;
        }

        // Find all build directories in examples
        for entry in WalkDir::new(&examples_dir)
            .into_iter()
            .filter_map(|e| e.ok())
        {
            if entry.file_type().is_dir() {
                let dir_name = entry.file_name().to_string_lossy();
                if dir_name == "build" || dir_name == ".gradle" || dir_name == ".hvigor" {
                    let rel_path = entry
                        .path()
                        .strip_prefix(&self.project_dir)
                        .unwrap()
                        .to_string_lossy()
                        .to_string();
                    self.remove_directory(entry.path(), Some(&rel_path));
                }
            }
        }
    }

    fn clean_platform(&mut self, platform: &str) {
        match platform {
            "android" => self.clean_android(),
            "ios" => self.clean_ios(),
            "macos" => self.clean_macos(),
            "ohos" => self.clean_ohos(),
            "kmp" => self.clean_kmp(),
            "examples" => self.clean_examples(),
            _ => {
                eprintln!("Unknown platform: {}", platform);
            }
        }
    }

    fn clean_all(&mut self) {
        println!("\n{}", "=".repeat(60));
        println!("  Cleaning ALL build artifacts and caches");
        println!("{}", "=".repeat(60));

        if !self.dry_run && !self.skip_confirm {
            if !self.confirm_clean("\n⚠️  This will remove ALL build artifacts. Continue?") {
                println!("  ⏭️  Aborted");
                return;
            }
        }

        // Clean in order
        self.clean_bin();
        self.clean_cmake();
        self.clean_android();
        self.clean_ios();
        self.clean_macos();
        self.clean_ohos();
        self.clean_kmp();
        self.clean_examples();

        // Also clean root .gradle if exists
        let root_gradle = self.project_dir.join(".gradle");
        if root_gradle.exists() {
            println!("\n{}", "=".repeat(60));
            println!("  Cleaning root .gradle/");
            println!("{}", "=".repeat(60));
            self.remove_directory(&root_gradle, Some(".gradle/"));
        }
    }

    fn print_summary(&self) {
        println!("\n{}", "=".repeat(60));
        println!("  Cleaning Summary");
        println!("{}", "=".repeat(60));

        if self.dry_run {
            println!("  [DRY RUN MODE - No files were actually deleted]");
        }

        if !self.cleaned_dirs.is_empty() {
            println!("  ✅ Successfully cleaned {} directories:", self.cleaned_dirs.len());
            for dir_name in &self.cleaned_dirs {
                println!("     - {}", dir_name);
            }
            println!("\n  💾 Total space freed: {}", Self::format_size(self.cleaned_size));
        } else {
            println!("  ℹ️  No directories were cleaned");
        }

        if !self.failed_dirs.is_empty() {
            println!("\n  ❌ Failed to clean {} directories:", self.failed_dirs.len());
            for (dir_name, error) in &self.failed_dirs {
                println!("     - {}: {}", dir_name, error);
            }
        }

        println!("{}\n", "=".repeat(60));

        if self.dry_run {
            println!("💡 Tip: Run without --dry-run to actually delete the files");
        }
    }
}

impl CleanCommand {
    /// Execute the clean command
    pub fn execute(self, _verbose: bool) -> Result<()> {
        println!("Cleaning build artifacts and caches...\n");

        // Get current working directory (project directory)
        let project_dir = std::env::current_dir()
            .context("Failed to get current working directory")?;

        // Find CCGO.toml to locate project directory
        let project_subdir = Self::find_project_dir(&project_dir)?;

        let mut cleaner = ProjectCleaner::new(project_subdir, self.dry_run, self.yes);

        // Determine what to clean based on arguments
        if self.native_only {
            // Only clean cmake_build directory
            cleaner.clean_cmake();
        } else if self.target == "all" {
            // Clean all platforms
            cleaner.clean_all();
        } else {
            // Clean specific platform
            cleaner.clean_platform(&self.target);
        }

        cleaner.print_summary();

        Ok(())
    }

    fn find_project_dir(current_dir: &Path) -> Result<PathBuf> {
        // Check if CCGO.toml exists in subdirectories
        if let Ok(entries) = fs::read_dir(current_dir) {
            for entry in entries.flatten() {
                if let Ok(file_type) = entry.file_type() {
                    if file_type.is_dir() {
                        let potential_config = entry.path().join("CCGO.toml");
                        if potential_config.exists() {
                            return Ok(entry.path());
                        }
                    }
                }
            }
        }

        // If not found in subdirectory, check current directory
        let config_path = current_dir.join("CCGO.toml");
        if config_path.exists() {
            return Ok(current_dir.to_path_buf());
        }

        // Warning if CCGO.toml not found
        eprintln!("⚠️  Warning: CCGO.toml not found. May not be in a CCGO project directory.");
        eprintln!("   Continuing anyway...\n");

        Ok(current_dir.to_path_buf())
    }
}
